spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java

```diff
@@ -0,0 +1,185 @@
+import org.springframework.ws.transport.TransportConstants;
+import org.springframework.ws.transport.WebServiceConnection;
+import org.springframework.ws.transport.context.TransportContext;
+import org.springframework.ws.transport.context.TransportContextHolder;
+import org.w3c.dom.Node;
+import org.xml.sax.SAXException;
+
+import javax.xml.namespace.QName;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import javax.xml.transform.Source;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.sax.SAXSource;
+import javax.xml.transform.stream.StreamSource;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Interceptor that creates an Observation for each operation.
+ *
+ * @author Johan Kindgren
+ * @see Observation
+ * @see io.micrometer.observation.ObservationConvention
+ */
+public class WebServiceObservationInterceptor extends ClientInterceptorAdapter {
+
+
+    public static final String OBSERVATION_KEY = "observation";
+    private ObservationRegistry observationRegistry;
+
+    private static final WebServiceTemplateConvention DEFAULT_CONVENTION = new DefaultWebServiceTemplateConvention();
+
+    private SAXParserFactory parserFactory;
+    private SAXParser saxParser;
+
+    private WebServiceTemplateConvention customConvention;
+
+    public WebServiceObservationInterceptor(ObservationRegistry observationRegistry) {
+        this.observationRegistry = observationRegistry;
+        parserFactory = SAXParserFactory.newNSInstance();
+        try {
+            saxParser = parserFactory.newSAXParser();
+        } catch (ParserConfigurationException e) {
+            throw new RuntimeException(e);
+        } catch (SAXException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+
+    @Override
+    public boolean handleRequest(MessageContext messageContext) throws WebServiceClientException {
+
+        TransportContext transportContext = TransportContextHolder.getTransportContext();
+        HeadersAwareSenderWebServiceConnection connection =
+                (HeadersAwareSenderWebServiceConnection) transportContext.getConnection();
+
+
+        Observation observation = WebServiceTemplateObservationDocumentation.WEB_SERVICE_TEMPLATE.start(
+                customConvention,
+                DEFAULT_CONVENTION,
+                () -> new WebServiceTemplateObservationContext(connection),
+                observationRegistry);
+
+        messageContext.setProperty(OBSERVATION_KEY, observation);
+
+        return true;
+    }
+
+    @Override
+    public void afterCompletion(MessageContext messageContext, Exception ex) throws WebServiceClientException {
+
+        Observation observation = (Observation) messageContext.getProperty(OBSERVATION_KEY);
+        Assert.notNull(observation, "Expected observation in messageContext");
```

Comment:
What's the effect of throwing an exception here? Will it result in an exception thrown from the call to WebServiceTemplate? I would generally advise against doing that. If the instrumentation is broken, we do want some way to notice that (logging perhaps), but user code that works without observability instrumentation should not fail when observability instrumentation is added. An exception to this rule of thumb might be for configuration issues where an exception on startup can be thrown, but this will be thrown when handling a request, if I understand correctly.

Reply:
I agree with your remark, the throws-statement is just a leftover from the generated method and should of course be removed if it isn't used.

Reply:
I meant the call to `Assert.nonNull` will throw an exception if `observation` is null, and we probably don't want that.

Reply:
Could this be a candidate for WarnThenDebugLogger, or should it just silently be ignored?

Reply:
Replaced the assertion with WarnThenDebugLogger